### DISCUSSION TITLE
How to get thread number with parallel execution for Allure report timeline?

### DISCUSSION THREAD

**Discussion by theptyza** [codeceptjs/CodeceptJS#3031] [Q&A]
#### What are you trying to achieve?
I'm running tests in parallel with 4 chunks and write reports to Allure.
I want to see tests from each chunk displayed in a separate timeline. If I manually add a "thread" label to each suite or test (in the XML) I can split the test results into multiple timelines.
![parallel timelines](https://user-images.githubusercontent.com/5357603/63260309-3bbc1200-c289-11e9-9c3e-c1fcb0023480.PNG)

#### What do you get instead?
By default in the Allure report all tests are displayed in one timeline.
![single timeline](https://user-images.githubusercontent.com/5357603/63260258-1929f900-c289-11e9-965b-80fed0aa374a.PNG)

I can add labels to test cases like
```
const allure = codeceptjs.container.plugins('allure');

Before(() => 
  allure.addLabel('thread', '...')
)
```

My question is - how can I get the **chunk number** in the test?

### Details

* CodeceptJS version: 2.3.0
* NodeJS Version: 10.16.3
* Operating System: Windows 10
* Configuration file:

```js
exports.config = {
  ...
  mocha: {
    reporterOptions: {
      reportDir: "./test/report",
      autoOpen: true,
    }
  },
  plugins: {
    allure: {
      enabled: true,
      outputDir: "./test/.output/allure-results"
    },
    stepByStepReport: {}
  },
  multiple: {
    parallel: {
      chunks: "4"
    }
  },
  bootstrap: false,
  teardown: null,
  hooks: [],
  tests: "./test/tests/ui/dashboard/*.js",
  timeout: 120000,
};
```

Running tests with 

```
npx codeceptjs run-multiple parallel
```

**Comment by xMutaGenx** [comment]
You can retrieve your chunks configuration with the following `const chunks = require('codeceptjs').config.get()['multiple']['parallel']['chunks'];` alternatively you can pass in the number of chunks to the node env, which then allows you to access that globally `CHUNKS=4 npx codeceptjs run-multiple parallel` then you can set them in your config via:
```
multiple: {
    parallel: {
      chunks: process.env.CHUNKS
    }
  }
```
As well as access them globally:
```
const allure = codeceptjs.container.plugins('allure');

Before(() => 
  allure.addLabel('thread', process.env.CHUNKS)
)
```

Hope this helps.

**Comment by theptyza** [comment]
@xMutaGenx thanks. What I'm trying to do is get the **current chunk number**.
So that tests from chunk 1 are recorded as thread 1, tests from chunk 2 as thread 2 etc.

With
```js
const chunks = require('codeceptjs').config.get()['multiple']['parallel']['chunks'];

Before(() => 
  allure.addLabel('thread', process.env.CHUNKS)
)
```

Every test is recorded as thread 4 (the number of chunks).

**Comment by xMutaGenx** [comment]
Ah apologies @theptyza misunderstood you then, one way you could do this is by accessing the mocha child processes:
```js
let child = require('codeceptjs').config.get().mocha.child;
```
This will give you something like `1.chrome:chunk1:chrome`
Next you'll want to get the number of your child process from the string:
```js
child = child.substring(0, child.indexOf('.'));
```
Finally you'll have something like:
```js
Before(() => {
    let child = require('codeceptjs').config.get().mocha.child;
    child = child.substring(0, child.indexOf('.'));
    allure.addLabel('thread', child)
})
```

Not the cleanest approach, but it works, might allow you to continue working until a better solution appears 🤷‍♂ Hope this helps.

**Comment by theptyza** [comment]
I was able to use process ID as thread number for Allure report by creating a helper that sets thread label for each test
```js
let Helper = codecept_helper;

const process = require('process');

class AllureHelper extends Helper {
    _before() {
        const allure = codeceptjs.container.plugins('allure');
        allure.addLabel('thread', process.pid);
    }
}

module.exports = AllureHelper;
```
![parallel pid](https://user-images.githubusercontent.com/5357603/63349793-d2a9cc80-c364-11e9-8004-af7a7ff788a7.png)

Wondering if there is a better way to do this.

**Comment by theptyza** [comment]
@xMutaGenx thanks! Your solution is even better since it gives the actual chunk number.

**Comment by Naimadnap** [comment]
@theptyza or @xMutaGenx could you share the configuration to achive this

**Comment by theptyza** [comment]
@Naimadnap I've created an allureHelper
```js
'use strict';

let Helper = codecept_helper;

class AllureHelper extends Helper {
    _before() {        
        let child = codeceptjs.config.get().mocha.child;
        // if there are child processes set thread label in Allure report to chunk number
        // child process name is like '1.parallel:chunk1:default'
        if(child) {
            const allure = codeceptjs.container.plugins('allure');
            allure.addLabel('thread', child.substring(0, child.indexOf('.')));
        }
    }
}

module.exports = AllureHelper;
```
and included it in codecept.conf.js
```js
helpers: {
    Nightmare: {
      url: process.env.CODECEPT_URL || "http://localhost:3000",
      windowSize: "1920x1080",
      typeInterval: "-1"
    },
    allureHelper: {
      require: "./test/helpers/allureHelper.js",
    }
  },
```

**Comment by jancorvus** [comment]
What about
```js
const { threadId } = require('worker_threads');
const allure = codeceptjs.container.plugins('allure');

allure.addLabel('thread', threadId);
```
?
